draft-ietf-quic-version-negotiation.md

```diff
@@ -67,105 +67,171 @@ Discussion of this work is encouraged to happen on the QUIC IETF mailing list
 <https://github.com/quicwg/version-negotiation/>.
 
 
-# Conventions and Definitions
+## Conventions and Definitions
 
 The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
 "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this
 document are to be interpreted as described in BCP 14 {{!RFC2119}} {{!RFC8174}}
 when, and only when, they appear in all capitals, as shown here.
 
 
+# Connections and Version Negotiation
+
+QUIC connections are shared state between a client and a server {{INV}}. The VN
+mechanism defined in this document operates inside of a QUIC connection; i.e.,
+the packets with the original version are part of the same connection as the VN
+packet and the packets with the negotiated version.
+
+
+# Compatible Versions
+
+If A and B are two distinct versions of QUIC, A is said to be "compatible" with
+B if it is possible to take a first flight of packets from version A and
+convert it into a first flight of packets from version B. As an example, if
+versions A and B are absolutely equal in their wire image and behavior during
+the handshake but differ after the handshake, then A is compatible with B and B
+is compatible with A.
+
+Version compatibility is not bijective: it is possible for version A to be
+compatible with version B and for B not to be compatible with A. This could
+happen for example if version B is a strict superset of version A.
+
+Note that version compatibility does not mean that every single possible
+instance of a first flight will succeed in conversion to the other version. For
+example, if version B is equal to A in all aspects except it introduced a new
+frame in its first flight that version A cannot parse or even ignore, then A
+could still be compatible with B as conversions would succeed for connections
+where that frame is not used.
+
+When a new version of QUIC is defined, it is assumed to not be compatible with
+any other version unless otherwise specified. Similarly, no other version is
+compatible with the new version unless otherwise specified. Implementations
+MUST NOT assume compatibility between versions unless explicitly specified.
+
+Note that both endpoints might disagree on whether two versions are compatible
+or not. For example, two versions could have been defined concurrently and then
+specified as compatible in a third document much later - in that scenario one
+endpoint might be aware of the compatibility document while the other may not.
+
```

Comment:
Trying to define this as a property of *versions* seems problematic. I think we would be better off with a much simpler model, focuses on flights. A version A first flight is compatible with B if it would be acceptable to a conforming B implementation (this is the situation in TLS, for instance). 

Consider the case where we start with version X.
We now define version Y which has the same initial flight as X except that it requires transport parameter A and requires an initial version of X or Y.

Thus, a version X Initial could consist of:
{version = X, compatible_versions = [X], transport_parameters = [...]}

A compatible version Y would consist of:

At some point in the future, when version X is deprecated, you could send:
{version = Y, compatible_versions = [X, Y], transport_parameters = [...., A]}

This doesn't preclude defining concurrent versions, in that one could could define a version Z which is like X but needs transport parameter B, so it has an X compatible first flight
{version = X, compatible_versions = [X, Z], transport_parameters = [...., B]}

It also has an X, Y, Z compatible first flight:
{version = X, compatible_versions = [X, Y, Z], transport_parameters = [...., A, B]}

This obviously works more cleanly if versions are essentially linear, but works OK if each version is required to accept an Initial which has the external version matching some previous version as well.